File: src/core/x3dhSession.js
// X3DH session orchestration — glue between the pure X3DH math, the local
// prekey store, and the cached remote bundles.
//
// Two entry points:
//   * `deriveInitiatorBootstrap(peerId)` — called by "alice" (lexicographically
//     smaller peerId) when she has a cached verified bundle for `peerId`.
//     Generates a fresh ephemeral EK, runs DH1..DH4, and returns both the SK
//     used to seed the Double Ratchet and the fields that must ride on the
//     outgoing hello_v4 so bob can replay the math.
//
//   * `deriveResponderBootstrap(fields)` — called by "bob" when an incoming
//     hello_v4 carries X3DH fields. Looks up his SPK priv by id, optionally
//     consumes the referenced OPK, verifies alice's x3dh-identity binding,
//     and derives the matching SK.
//
// On a successful initiator bootstrap we delete the cached bundle so the next
// session refetches a fresh OPK — reusing an OPK across sessions would break
// X3DH forward secrecy (bob's consume-once check would also reject it).

import { getCachedBundle, deleteCachedBundle } from './bundleCache.js';
import {
  getOrCreateX3DHIdentity,
  exportX3DHIdentityPubSpki,
  verifyX3DHBinding
} from './identityKey.js';
import { getSignedPrekeyById, consumeOPK } from './prekeyStore.js';
import {
  initiatorX3DH,
  responderX3DH,
  generateEphemeralECDHPair,
  exportECDHPubSpki
} from './x3dh.js';

/**
 * Attempt to bootstrap X3DH as the initiator for `peerId`. Returns `null` if
 * we have no cached bundle (caller should fall back to the plain DH hello).
 * On success:
 *   {
 *     sk,                  // Uint8Array(32) — seed for the Double Ratchet root key
 *     ekSpki,              // SPKI bytes of the fresh ephemeral ECDH pub
 *     spkId,               // bob's SPK id we DH'd against
 *     opkId,               // bob's OPK id we consumed, or null
 *     myX3dhIkSpki,        // our long-term X3DH ECDH pub SPKI
 *     myX3dhIkSig          // our ECDSA binding over myX3dhIkSpki
 *   }
 */
export async function deriveInitiatorBootstrap(peerId) {
  const cached = await getCachedBundle(peerId);
  if (!cached) return null;
  const { bundle } = cached;

  const myIdentity = await getOrCreateX3DHIdentity();
  const myX3dhIkSpki = await exportX3DHIdentityPubSpki();

  const ek = await generateEphemeralECDHPair();
  const ekSpki = await exportECDHPubSpki(ek.publicKey);

  const { sk } = await initiatorX3DH({
    IK_a_priv: myIdentity.privateKey,
    EK_a_priv: ek.privateKey,
    IK_b_spki: bundle.x3dhIdentitySpki,
    SPK_b_spki: bundle.spk.pub,
    OPK_b_spki: bundle.opk ? bundle.opk.pub : null
  });

  // One-shot: drop the cache so the next connection pulls a fresh bundle
  // (bob's OPK is now consumed on his side once he processes our hello).
  try { await deleteCachedBundle(peerId); } catch (_) {}

  return {
    sk,
    ekSpki,
    spkId: bundle.spk.id,
    opkId: bundle.opk ? bundle.opk.id : null,
    myX3dhIkSpki,
    myX3dhIkSig: myIdentity.bindingSig
  };
}

/**
 * Replay the X3DH DHs on bob's side. Returns `{ ok:true, sk }` on success or
 * `{ ok:false, reason }` on any failure (unknown SPK, already-consumed OPK,
 * malformed binding sig, etc.). Caller decides whether to abort the handshake
 * or fall back — in our wiring we hard-fail, since silently dropping to plain
 * DH would leave alice using a different bootstrap than bob.
 */
export async function deriveResponderBootstrap({
  senderIdSpki,
  senderX3dhIkSpki,
  senderX3dhIkSig,
  ekSpki,
  spkId,
  opkId = null
}) {
  if (!senderIdSpki || !senderX3dhIkSpki || !senderX3dhIkSig) {
    return { ok: false, reason: 'x3dh: missing sender identity fields' };
  }
  if (!ekSpki || !spkId) {
    return { ok: false, reason: 'x3dh: missing ek or spkId' };
  }

  const bindingOk = await verifyX3DHBinding(senderIdSpki, senderX3dhIkSpki, senderX3dhIkSig);
  if (!bindingOk) return { ok: false, reason: 'x3dh: sender binding invalid' };

  const spk = await getSignedPrekeyById(spkId);
  if (!spk) return { ok: false, reason: `x3dh: unknown spk ${spkId}` };

  let opkPriv = null;
  if (opkId) {
    const opk = await consumeOPK(opkId);
    if (!opk) return { ok: false, reason: `x3dh: unknown or consumed opk ${opkId}` };
    opkPriv = opk.privateKey;
  }

  const myIdentity = await getOrCreateX3DHIdentity();
  const { sk } = await responderX3DH({
    SPK_b_priv: spk.privateKey,
    IK_b_priv: myIdentity.privateKey,
    OPK_b_priv: opkPriv,
    IK_a_spki: senderX3dhIkSpki,
    EK_a_spki: ekSpki
  });
  return { ok: true, sk };
}

File: lib/core/x3dh_session.dart
// Port of src/core/x3dhSession.js — X3DH orchestration.
//
// Glue between the pure X3DH math (x3dh.dart), the local prekey store, and
// the cached remote bundles. Two entry points:
//
//   * [deriveInitiatorBootstrap] — alice side. Needs a cached verified bundle
//     for [peerId]. Mints EK, runs DH1..DH4, returns SK plus the fields that
//     must ride on the outgoing signed-hello v4 so bob can replay the math.
//
//   * [deriveResponderBootstrap] — bob side. Looks up his SPK priv by id,
//     optionally consumes the referenced OPK, verifies alice's x3dh-identity
//     binding, and derives the matching SK.
//
// On a successful initiator bootstrap we drop the cached bundle: reusing an
// OPK across sessions breaks forward secrecy, and bob's consume-once check
// would reject it anyway.

import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';

import 'bundle_cache.dart';
import 'identity_key.dart' as identity_key;
import 'prekey_store.dart';
import 'x3dh.dart';

/// Fields the initiator must put on the wire so the responder can run the
/// matching X3DH. Matches the JS return shape of [deriveInitiatorBootstrap].
class InitiatorBootstrap {
  const InitiatorBootstrap({
    required this.sk,
    required this.ekSpki,
    required this.spkId,
    required this.myX3dhIkSpki,
    required this.myX3dhIkSig,
    this.opkId,
  });

  /// 32-byte session seed for the Double Ratchet root key.
  final Uint8List sk;

  /// SPKI bytes of the fresh ephemeral ECDH pub.
  final Uint8List ekSpki;

  /// The responder's SPK id we DH'd against.
  final String spkId;

  /// The responder's OPK id we consumed, or null if none was published.
  final String? opkId;

  /// Our long-term X3DH ECDH pub SPKI.
  final Uint8List myX3dhIkSpki;

  /// Our ECDSA binding signature over [myX3dhIkSpki].
  final Uint8List myX3dhIkSig;
}

/// Attempt to bootstrap X3DH as the initiator for [peerId]. Returns null if
/// there's no cached bundle — caller should fall back to the plain DH hello.
Future<InitiatorBootstrap?> deriveInitiatorBootstrap(String peerId) async {
  final cached = await getCachedBundle(peerId);
  if (cached == null) return null;
  final bundle = cached.bundle;

  final myIdentity = await identity_key.getOrCreateX3DHIdentity();
  final myX3dhIkSpki = await identity_key.exportX3DHIdentityPubSpki();

  final ek = await generateEphemeralECDHPair();
  final ekSpki = await exportECDHPubSpki(ek);

  final result = await initiatorX3DH(
    ikAPriv: myIdentity.keyPair,
    ekAPriv: ek,
    ikBSpki: bundle.x3dhIdentitySpki,
    spkBSpki: bundle.spk.pub,
    opkBSpki: bundle.opk?.pub,
  );

  // One-shot: drop the cached bundle so the next connection pulls a fresh one
  // (bob's OPK is now consumed on his side once he processes our hello).
  try {
    await deleteCachedBundle(peerId);
  } catch (_) {}

  return InitiatorBootstrap(
    sk: result.sk,
    ekSpki: ekSpki,
    spkId: bundle.spk.id,
    opkId: bundle.opk?.id,
    myX3dhIkSpki: myX3dhIkSpki,
    myX3dhIkSig: myIdentity.bindingSig,
  );
}

/// Outcome of a responder-side bootstrap.
class ResponderBootstrap {
  const ResponderBootstrap({required this.ok, this.sk, this.reason});
  final bool ok;

  /// 32-byte session seed — present iff [ok] is true.
  final Uint8List? sk;
  final String? reason;
}

/// Replay the X3DH DHs on bob's side. Returns `ok=false` with a reason on any
/// failure (unknown SPK, already-consumed OPK, malformed binding sig, etc.)
/// so the caller can hard-fail the handshake — silently falling back to plain
/// DH would leave alice using a different bootstrap than bob.
Future<ResponderBootstrap> deriveResponderBootstrap({
  required List<int>? senderIdSpki,
  required List<int>? senderX3dhIkSpki,
  required List<int>? senderX3dhIkSig,
  required List<int>? ekSpki,
  required String? spkId,
  String? opkId,
}) async {
  if (senderIdSpki == null ||
      senderX3dhIkSpki == null ||
      senderX3dhIkSig == null) {
    return const ResponderBootstrap(
      ok: false,
      reason: 'x3dh: missing sender identity fields',
    );
  }
  if (ekSpki == null || spkId == null || spkId.isEmpty) {
    return const ResponderBootstrap(
      ok: false,
      reason: 'x3dh: missing ek or spkId',
    );
  }

  final bindingOk = await identity_key.verifyX3DHBinding(
    senderIdSpki,
    senderX3dhIkSpki,
    senderX3dhIkSig,
  );
  if (!bindingOk) {
    return const ResponderBootstrap(
      ok: false,
      reason: 'x3dh: sender binding invalid',
    );
  }

  final spk = await getSignedPrekeyById(spkId);
  if (spk == null) {
    return ResponderBootstrap(ok: false, reason: 'x3dh: unknown spk $spkId');
  }

  EcKeyPair? opkPriv;
  if (opkId != null && opkId.isNotEmpty) {
    final opk = await consumeOPK(opkId);
    if (opk == null) {
      return ResponderBootstrap(
        ok: false,
        reason: 'x3dh: unknown or consumed opk $opkId',
      );
    }
    opkPriv = opk.privateKey;
  }

  final myIdentity = await identity_key.getOrCreateX3DHIdentity();
  final result = await responderX3DH(
    spkBPriv: spk.privateKey,
    ikBPriv: myIdentity.keyPair,
    opkBPriv: opkPriv,
    ikASpki: senderX3dhIkSpki,
    ekASpki: ekSpki,
  );
  return ResponderBootstrap(ok: true, sk: result.sk);
}
